File: blueprints/clean.py
import os
import json
import pandas as pd
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required
from blueprints.utils import get_user_session
from backend.file_handler import load_df, save_df, log_operation, get_ops_log, reapply_ops
from backend import clean_data

clean_bp = Blueprint('clean', __name__, url_prefix='/api')

def session_guard():
    session_folder, filename, user_id, _ = get_user_session()
    if not session_folder:
        return None, None, jsonify({'error': 'No file uploaded'}), 404
    return session_folder, filename, None, None


@clean_bp.route('/clean/columns', methods=['GET'])
@jwt_required()
def get_columns():
    session_folder, filename, user_id, _ = get_user_session()
    if not session_folder:
        return jsonify({'error': 'No file uploaded'}), 404
    df = load_df(session_folder)
    columns = [{'name': col, 'type': str(df[col].dtype)} for col in df.columns]
    return jsonify({'columns': columns})


@clean_bp.route('/clean/null', methods=['POST'])
@jwt_required()
def clean_null():
    session_folder, _, __, ___ = get_user_session()
    if not session_folder: return jsonify({'error': 'No file uploaded'}), 404
    data = request.get_json()
    df = load_df(session_folder)
    result = clean_data.clean_null(df, data['column'], data['method'], data.get('custom_val', ''))
    if result['status'] == 'success':
        save_df(result['df'], session_folder)
        log_operation(session_folder, {'action': 'clean_null', 'column': data['column'], 'method': data['method'], 'custom_val': data.get('custom_val', '')})
        return jsonify({'success': True})
    return jsonify({'error': result['error']}), 400


@clean_bp.route('/clean/drop-rows', methods=['POST'])
@jwt_required()
def drop_rows():
    session_folder, _, __, ___ = get_user_session()
    if not session_folder: return jsonify({'error': 'No file uploaded'}), 404
    data = request.get_json()
    df = load_df(session_folder)
    result = clean_data.drop_rows(df, data['scope'], data.get('column', ''))
    if result['status'] == 'success':
        save_df(result['df'], session_folder)
        log_operation(session_folder, {'action': 'drop_rows', 'scope': data['scope'], 'column': data.get('column', '')})
        return jsonify({'success': True})
    return jsonify({'error': result['error']}), 400


@clean_bp.route('/clean/drop-column', methods=['POST'])
@jwt_required()
def drop_column():
    session_folder, _, __, ___ = get_user_session()
    if not session_folder: return jsonify({'error': 'No file uploaded'}), 404
    data = request.get_json()
    df = load_df(session_folder)
    result = clean_data.drop_column(df, data['column'])
    if result['status'] == 'success':
        save_df(result['df'], session_folder)
        log_operation(session_folder, {'action': 'drop_column', 'column': data['column']})
        return jsonify({'success': True})
    return jsonify({'error': result['error']}), 400


@clean_bp.route('/clean/rename-column', methods=['POST'])
@jwt_required()
def rename_column():
    session_folder, _, __, ___ = get_user_session()
    if not session_folder: return jsonify({'error': 'No file uploaded'}), 404
    data = request.get_json()
    df = load_df(session_folder)
    result = clean_data.rename_column(df, data['column'], data['new_name'])
    if result['status'] == 'success':
        save_df(result['df'], session_folder)
        log_operation(session_folder, {'action': 'rename_column', 'column': data['column'], 'new_name': data['new_name']})
        return jsonify({'success': True})
    return jsonify({'error': result['error']}), 400


@clean_bp.route('/clean/remove-duplicates', methods=['POST'])
@jwt_required()
def remove_duplicates():
    session_folder, _, __, ___ = get_user_session()
    if not session_folder: return jsonify({'error': 'No file uploaded'}), 404
    data = request.get_json()
    df = load_df(session_folder)
    result = clean_data.remove_duplicates(df, data['keep'])
    if result['status'] == 'success':
        save_df(result['df'], session_folder)
        log_operation(session_folder, {'action': 'remove_duplicates', 'keep': data['keep']})
        return jsonify({'success': True})
    return jsonify({'error': result['error']}), 400


@clean_bp.route('/clean/change-dtype', methods=['POST'])
@jwt_required()
def change_dtype():
    session_folder, _, __, ___ = get_user_session()
    if not session_folder: return jsonify({'error': 'No file uploaded'}), 404
    data = request.get_json()
    df = load_df(session_folder)
    result = clean_data.change_dtype(df, data['column'], data['type_'])
    if result['status'] == 'success':
        save_df(result['df'], session_folder)
        log_operation(session_folder, {'action': 'change_dtype', 'column': data['column'], 'type_': data['type_']})
        return jsonify({'success': True})
    return jsonify({'error': result['error']}), 400


@clean_bp.route('/clean/replace-val', methods=['POST'])
@jwt_required()
def replace_val():
    session_folder, _, __, ___ = get_user_session()
    if not session_folder: return jsonify({'error': 'No file uploaded'}), 404
    data = request.get_json()
    df = load_df(session_folder)
    result = clean_data.replace_val(df, data['column'], data['find_val'], data['replace_val'], data['exact_match'])
    if result['status'] == 'success':
        save_df(result['df'], session_folder)
        log_operation(session_folder, {'action': 'replace_val', 'column': data['column'], 'find_val': data['find_val'], 'replace_val': data['replace_val'], 'exact_match': data['exact_match']})
        return jsonify({'success': True})
    return jsonify({'error': result['error']}), 400


@clean_bp.route('/clean/trim-space', methods=['POST'])
@jwt_required()
def trim_space():
    session_folder, _, __, ___ = get_user_session()
    if not session_folder: return jsonify({'error': 'No file uploaded'}), 404
    data = request.get_json()
    df = load_df(session_folder)
    result = clean_data.trim_space(df, data['column'])
    if result['status'] == 'success':
        save_df(result['df'], session_folder)
        log_operation(session_folder, {'action': 'trim_space', 'column': data['column']})
        return jsonify({'success': True})
    return jsonify({'error': result['error']}), 400


@clean_bp.route('/clean/change-case', methods=['POST'])
@jwt_required()
def change_case():
    session_folder, _, __, ___ = get_user_session()
    if not session_folder: return jsonify({'error': 'No file uploaded'}), 404
    data = request.get_json()
    df = load_df(session_folder)
    result = clean_data.change_case(df, data['column'], data['case'])
    if result['status'] == 'success':
        save_df(result['df'], session_folder)
        log_operation(session_folder, {'action': 'change_case', 'column': data['column'], 'case': data['case']})
        return jsonify({'success': True})
    return jsonify({'error': result['error']}), 400


@clean_bp.route('/clean/reorder-columns', methods=['POST'])
@jwt_required()
def reorder_columns():
    session_folder, _, __, ___ = get_user_session()
    if not session_folder: return jsonify({'error': 'No file uploaded'}), 404
    data = request.get_json()
    df = load_df(session_folder)
    result = clean_data.reorder_columns(df, data['new_order'])
    if result['status'] == 'success':
        save_df(result['df'], session_folder)
        log_operation(session_folder, {'action': 'reorder_columns', 'new_order': data['new_order']})
        return jsonify({'success': True})
    return jsonify({'error': result['error']}), 400


@clean_bp.route('/clean/undo', methods=['POST'])
@jwt_required()
def undo():
    session_folder, _, __, ___ = get_user_session()
    if not session_folder: return jsonify({'error': 'No file uploaded'}), 404
    log = get_ops_log(session_folder)
    if not log:
        return jsonify({'error': 'Nothing to undo'}), 400
    log.pop()
    log_path = os.path.join(session_folder, 'ops_log.json')
    with open(log_path, 'w') as f:
        json.dump(log, f, indent=2)
    df = pd.read_csv(os.path.join(session_folder, 'original.csv'))
    df = reapply_ops(df, log)
    save_df(df, session_folder)
    return jsonify({'success': True})

File: static/js/clean.js
/* ── CLEAN.JS ─────────────────────────────────────
   All clean tab operations, sidebar, reorder panel
──────────────────────────────────────────────── */

// ── STATE ────────────────────────────────────────
let selectedCol  = null;
let selectedType = null;

// Column type map — built from COLUMNS passed by Flask
const colTypeMap = {};
COLUMNS.forEach(col => { colTypeMap[col.name] = col.type; });

// ── DARK MODE ────────────────────────────────────
function initTheme() {
  const saved = localStorage.getItem('pilotdf-theme') || 'light';
  document.documentElement.setAttribute('data-theme', saved);
  document.getElementById('themeIcon').textContent = saved === 'dark' ? '☀️' : '🌙';
}
document.getElementById('themeToggle').addEventListener('click', () => {
  const current = document.documentElement.getAttribute('data-theme');
  const next = current === 'dark' ? 'light' : 'dark';
  document.documentElement.setAttribute('data-theme', next);
  localStorage.setItem('pilotdf-theme', next);
  document.getElementById('themeIcon').textContent = next === 'dark' ? '☀️' : '🌙';
});

// ── TOAST ────────────────────────────────────────
function showToast(msg, type = 'success') {
  const toast = document.getElementById('toast');
  toast.textContent = msg;
  toast.className = `toast ${type}`;
  setTimeout(() => { toast.className = 'toast'; }, 3000);
}

// ── API CALL HELPER ──────────────────────────────
async function callAPI(url, formData) {
  const res = await fetch(url, {
    method: 'POST',
    headers: { 'X-Requested-With': 'XMLHttpRequest' },
    body: formData
  });
  return await res.json();
}

function makeForm(obj) {
  const fd = new FormData();
  Object.entries(obj).forEach(([k, v]) => fd.append(k, v));
  return fd;
}

// ── SIDEBAR ──────────────────────────────────────
function buildSidebar(columns) {
  const list  = document.getElementById('colList');
  const count = document.getElementById('colCount');
  count.textContent = columns.length;
  list.innerHTML = '';

  columns.forEach(col => {
    const item = document.createElement('div');
    item.className = 'col-item';
    item.dataset.col  = col.name;
    item.dataset.type = col.type;
    item.innerHTML = `
      <span class="col-dot ${col.type}"></span>
      <span>${col.name}</span>
    `;
    item.addEventListener('click', () => selectColumn(col.name, col.type, item));
    list.appendChild(item);
  });

  buildReorderList(columns.map(c => c.name));
}

// Search filter
document.getElementById('colSearch').addEventListener('input', (e) => {
  const q = e.target.value.toLowerCase();
  document.querySelectorAll('.col-item').forEach(item => {
    item.style.display = item.dataset.col.toLowerCase().includes(q) ? '' : 'none';
  });
});

document.querySelector('.btn-reset').addEventListener('click', (e) => {
  if (!confirm('Reset will restore the original file and undo ALL operations. Are you sure?')) {
    e.preventDefault();
  }
});
// ── SELECT COLUMN ────────────────────────────────
function selectColumn(name, type, el) {
  selectedCol  = name;
  selectedType = type;

  document.querySelectorAll('.col-item').forEach(i => i.classList.remove('active'));
  el.classList.add('active');

  document.getElementById('emptyState').style.display = 'none';
  document.getElementById('opsPanel').style.display   = 'flex';
  const ops = document.getElementById('opsPanel');
  ops.style.display = 'flex';
  ops.style.animation = 'none';
  ops.offsetHeight; // force reflow
  ops.style.animation = '';
  const badge = document.getElementById('selectedColType');
  document.getElementById('selectedColName').textContent = name;
  badge.textContent = type;
  badge.className   = `col-type-badge ${type}`;

  document.getElementById('toast').className = 'toast';
  
  selectedCol  = name;
  selectedType = type;
  localStorage.setItem('pilotdf-clean-col', name);  // ← add this line
  document.querySelector('.main-panel').scrollTop = 0;
  // rest of your existing code...

}

// ── CHECK URL PARAM (from right-click in view tab) ──
function checkUrlParam() {
  const params = new URLSearchParams(window.location.search);
  const col    = params.get('col') || localStorage.getItem('pilotdf-clean-col');
  if (col) {
    const item = document.querySelector(`.col-item[data-col="${col}"]`);
    if (item) item.click();
  }
}

// ── FILL NULLS ───────────────────────────────────
document.getElementById('fillMethod').addEventListener('change', (e) => {
  document.getElementById('customValRow').style.display =
    e.target.value === 'custom' ? 'flex' : 'none';
});

document.getElementById('applyFillNull').addEventListener('click', async () => {
  if (!selectedCol) return;
  const method     = document.getElementById('fillMethod').value;
  const custom_val = document.getElementById('customVal').value;
  const data = await callAPI('/clean-clean_null', makeForm({ column: selectedCol, method, custom_val }));
  data.success ? showToast('Nulls filled successfully') : showToast(data.error, 'error');
});

// ── DROP ROWS ────────────────────────────────────
document.getElementById('applyDropRows').addEventListener('click', async () => {
  if (!selectedCol) return;
  const scope  = document.getElementById('dropScope').value;
  const column = scope === 'column' ? selectedCol : '';
  const data   = await callAPI('/clean-drop_rows', makeForm({ scope, column }));
  data.success ? showToast('Rows dropped successfully') : showToast(data.error, 'error');
});

// ── RENAME COLUMN ────────────────────────────────
document.getElementById('applyRename').addEventListener('click', async () => {
  if (!selectedCol) return;
  const new_name = document.getElementById('newColName').value.trim();
  if (!new_name) return showToast('Please enter a new name', 'error');
  const data = await callAPI('/clean-rename_column', makeForm({ column: selectedCol, new_name }));
  if (data.success) {
    showToast('Column renamed successfully');
    // update sidebar
    const item = document.querySelector(`.col-item[data-col="${selectedCol}"]`);
    if (item) { item.dataset.col = new_name; item.querySelector('span:last-child').textContent = new_name; }
    selectedCol = new_name;
    document.getElementById('selectedColName').textContent = new_name;
    document.getElementById('newColName').value = '';
  } else {
    showToast(data.error, 'error');
  }
});

// ── CHANGE DTYPE ─────────────────────────────────
document.getElementById('applyDtype').addEventListener('click', async () => {
  if (!selectedCol) return;
  const type_ = document.getElementById('newDtype').value;
  const data  = await callAPI('/clean-change_dtype', makeForm({ column: selectedCol, type_ }));
  data.success ? showToast('Data type changed successfully') : showToast(data.error, 'error');
});

// ── FIND & REPLACE ───────────────────────────────
document.getElementById('applyReplace').addEventListener('click', async () => {
  if (!selectedCol) return;
  const find_val    = document.getElementById('findVal').value;
  const replace_val = document.getElementById('replaceVal').value;
  const exact_match = document.getElementById('exactMatch').value;
  if (!find_val) return showToast('Please enter a value to find', 'error');
  const data = await callAPI('/clean-replace_val', makeForm({ column: selectedCol, find_val, replace_val, exact_match }));
  data.success ? showToast('Values replaced successfully') : showToast(data.error, 'error');
});

// ── TRIM WHITESPACE ──────────────────────────────
document.getElementById('applyTrim').addEventListener('click', async () => {
  if (!selectedCol) return;
  const data = await callAPI('/clean-trim_space', makeForm({ column: selectedCol }));
  data.success ? showToast('Whitespace trimmed successfully') : showToast(data.error, 'error');
});

// ── CHANGE CASE ──────────────────────────────────
document.getElementById('applyCase').addEventListener('click', async () => {
  if (!selectedCol) return;
  const case_ = document.getElementById('caseType').value;
  const data  = await callAPI('/clean-change_case', makeForm({ column: selectedCol, case: case_ }));
  data.success ? showToast('Case changed successfully') : showToast(data.error, 'error');
});

// ── DROP COLUMN ──────────────────────────────────
document.getElementById('applyDropCol').addEventListener('click', async () => {
  if (!selectedCol) return;
  if (!confirm(`Drop column "${selectedCol}"? This cannot be undone easily.`)) return;
  const data = await callAPI('/clean-drop_column', makeForm({ column: selectedCol }));
  if (data.success) {
    showToast('Column dropped successfully');
    const item = document.querySelector(`.col-item[data-col="${selectedCol}"]`);
    if (item) item.remove();
    selectedCol = null;
    document.getElementById('emptyState').style.display = 'flex';
    document.getElementById('opsPanel').style.display   = 'none';
  } else {
    showToast(data.error, 'error');
  }
});

// ── REMOVE DUPLICATES MODAL ──────────────────────
document.getElementById('removeDupBtn').addEventListener('click', () => {
  document.getElementById('dupModal').style.display = 'flex';
});
document.getElementById('cancelDup').addEventListener('click', () => {
  document.getElementById('dupModal').style.display = 'none';
});
document.getElementById('confirmDup').addEventListener('click', async () => {
  const keep = document.getElementById('dupKeep').value;
  const data = await callAPI('/clean-remove_duplicates', makeForm({ keep }));
  document.getElementById('dupModal').style.display = 'none';
  data.success ? showToast('Duplicates removed successfully') : showToast(data.error, 'error');
});

// ── DROP ALL NULL ROWS ───────────────────────────
document.getElementById('dropAllNullBtn').addEventListener('click', async () => {
  if (!confirm('Drop all rows with at least one null value?')) return;
  const data = await callAPI('/clean-drop_rows', makeForm({ scope: 'any', column: '' }));
  data.success ? showToast('Null rows dropped successfully') : showToast(data.error, 'error');
});

// ── UNDO ─────────────────────────────────────────
document.getElementById('undoBtn').addEventListener('click', async () => {
  const res  = await fetch('/clean-undo', {
    method: 'POST',
    headers: { 'X-Requested-With': 'XMLHttpRequest' }
  });
  const data = await res.json();
  data.success ? showToast('Last operation undone') : showToast(data.error || 'Nothing to undo', 'error');
  if (data.success) {
    showToast('Last operation undone');
    localStorage.removeItem('pilotdf-clean-col');  // clear so it doesn't try to restore
    setTimeout(() => window.location.reload(), 800);
}

});

// ── REORDER PANEL ────────────────────────────────
function buildReorderList(columns) {
  const list = document.getElementById('reorderList');
  list.innerHTML = '';
  columns.forEach(col => {
    const item = document.createElement('div');
    item.className   = 'reorder-item';
    item.draggable   = true;
    item.dataset.col = col;
    item.innerHTML   = `<span class="drag-handle">⠿</span><span>${col}</span>`;
    list.appendChild(item);
  });
  initDragDrop();
}

function initDragDrop() {
  const list  = document.getElementById('reorderList');
  let dragEl  = null;

  list.querySelectorAll('.reorder-item').forEach(item => {
    item.addEventListener('dragstart', () => { dragEl = item; item.classList.add('dragging'); });
    item.addEventListener('dragend',   () => { dragEl = null; item.classList.remove('dragging'); });
    item.addEventListener('dragover',  (e) => {
      e.preventDefault();
      if (dragEl && dragEl !== item) {
        list.querySelectorAll('.reorder-item').forEach(i => i.classList.remove('drag-over'));
        item.classList.add('drag-over');
        const rect = item.getBoundingClientRect();
        const mid  = rect.top + rect.height / 2;
        if (e.clientY < mid) list.insertBefore(dragEl, item);
        else list.insertBefore(dragEl, item.nextSibling);
      }
    });
    item.addEventListener('dragleave', () => item.classList.remove('drag-over'));
    item.addEventListener('drop', (e) => { e.preventDefault(); item.classList.remove('drag-over'); });
  });
}

document.getElementById('applyReorder').addEventListener('click', async () => {
  const items     = document.querySelectorAll('.reorder-item');
  const new_order = Array.from(items).map(i => i.dataset.col);
  const fd        = new FormData();
  new_order.forEach(col => fd.append('new_order', col));
  const res  = await fetch('/clean-reorder_columns', {
    method: 'POST',
    headers: { 'X-Requested-With': 'XMLHttpRequest' },
    body: fd
  });
  const data = await res.json();
  data.success ? showToast('Columns reordered successfully') : showToast(data.error, 'error');
});

// ── INIT ─────────────────────────────────────────
initTheme();
buildSidebar(COLUMNS);
checkUrlParam();